src/tasks/DockerComposeTaskProvider.ts

```diff
@@ -0,0 +1,85 @@
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import { cloneObject } from '../utils/cloneObject';
+import { CommandLineBuilder } from '../utils/commandLineBuilder';
+import { resolveVariables } from '../utils/resolveVariables';
+import { DockerComposeOptions, DockerComposeTaskDefinitionBase } from './DockerComposeTaskDefinitionBase';
+import { DockerTaskProvider } from './DockerTaskProvider';
+import { DockerComposeTaskContext, throwIfCancellationRequested } from './TaskHelper';
+
+// This doesn't need to be extended so redefining for parity is simplest
+export type DockerComposeTaskDefinition = DockerComposeTaskDefinitionBase;
+
+export interface DockerComposeTask extends Task {
+    definition: DockerComposeTaskDefinition;
+}
+
+export class DockerComposeTaskProvider extends DockerTaskProvider {
+    public constructor() { super('docker-build', undefined) }
+
+    protected async executeTaskInternal(context: DockerComposeTaskContext, task: DockerComposeTask): Promise<void> {
+        const definition = cloneObject(task.definition);
+        definition.dockerCompose = definition.dockerCompose || {};
+        definition.dockerCompose.files = definition.dockerCompose.files || [];
+
+        await this.validateResolvedDefinition(context, definition.dockerCompose);
+
+        const commandLine = await this.resolveCommandLine(definition.dockerCompose);
+
+        // Because BuildKit outputs everything to stderr, we will not treat output there as a failure
+        await context.terminal.executeCommandInTerminal(
+            commandLine,
+            context.folder,
+            false, // rejectOnStderr
+            undefined, // stdoutBuffer
+            Buffer.alloc(10 * 1024), // stderrBuffer
+            context.cancellationToken
+        );
+        throwIfCancellationRequested(context);
+    }
+
+    private async validateResolvedDefinition(context: DockerComposeTaskContext, dockerCompose: DockerComposeOptions): Promise<void> {
+        if (!dockerCompose.up && !dockerCompose.down) {
+            throw new Error(localize('vscode-docker.tasks.composeProvider.noUpOrDown', 'Neither the "up" or "down" properties are present in the docker-compose task.'));
+        }
+
+        for (const file of dockerCompose.files) {
+            if (!(await fse.pathExists(path.resolve(context.folder.uri.fsPath, resolveVariables(file, context.folder))))) {
+                throw new Error(localize('vscode-docker.tasks.composeProvider.invalidFile', 'One or more docker-compose files does not exist or could not be accessed.'));
+            }
+        }
+    }
+
+    private async resolveCommandLine(options: DockerComposeOptions): Promise<CommandLineBuilder> {
+        const isNewContext = isNewContextType((await ext.dockerContextManager.getCurrentContext()).ContextType);
+
+        if (options.up) {
+            return CommandLineBuilder
+                .create(isNewContext ? 'docker compose' : 'docker-compose')

```

Comment:
I wonder if we should move this to our Docker client API to consolidate it in one place.

Reply:
I agree. Since it depends on context, we should have it somewhere in there.